ships/0021-local-source-upload.md

```diff
@@ -0,0 +1,348 @@
+containers have a `/workspace/source` volume mounted, this volume will receive the uploaded data. The
+build-pod waits _(4)_ for the upload by keeping the entrypoint process active until the end of the
+upload.
+
+The CLI knows in which state _(5)_ the build-pod is at, since the data upload can only happen when
+the build-pod is running. Given the right conditions, it streams the local data to the cluster _(6)_.
+
+Once the streaming is finished, the CLI notifies the waiting process the upload is done _(7)_, so
+inside the build-pod the execution of subsequent containers can take place.
+
+#### Build _(Step 8 and 9)_
+
+When the build steps are being executed _(8)_ in the build-pod, the developer will have the execution
+logs being shown in the command-line _(9)_. The developer can interactively watch the container build
+execution and possible errors that might surface.
+
+At the end, when the build-pod is completed successfully _(5)_, the CLI shows the meta information
+about the container image built, and instructions on how to pull and use the image.
+
+#### Watching the Build-Pod
+
+As mentioned before, Shipwright builds are based on a sequence of containers being executed on a
+single pod, the CLI needs to watch _(5)_ over the state changes, and react as soon as it happens. For
+example, the build-pod can be deleted, fail immediately, fail during execution, and such. Moreover,
+as we can observe on the sequence diagram, it also needs to wait for specific conditions before
+streaming data.
+
+During research and development, this requirement was translated into a [watch][ClientGoWatch]
+instantiated on the build-pod, and a straightforward mechanism to create an event-loop around pod
+[state changes][EventLoopExample].  Each change of state will trigger the call of the informed
+closures, and allows wiring [events and reactions in a practical manner][EventLoopUsage].
+
+### Upload Mechanism
+
+This EP describes the direct data streaming to the Kubernetes cluster, directly to the POD running
+the actual build process, while we already have plans to support data upload towards a
+[container-registry][LocalSourceCodeEP] alternatively.
+
+#### Redirecting Standard Input
+
+The data upload mechanism is the same employed on <code>[kubectl cp][KubectlCP]</code>, which is
+redirecting the standard-input (<code>STDIN</code>) from the client to the container, inside the
+desired pod.
+
+Analogous to `kubectl cp`, we will be using `tar` to stream data from the client, while
+<code>[kubectl exec][KubectlExec]</code> on the pod keeps a <code>tar</code> process waiting for
+data on <code>STDIN</code>.
+
+#### Shipwright Build Controller
+
+The Shipwright Build Controller works based on [Tekton Workspaces][TektonWorkspaces], and therefore,
+it already defines the /workspace/source mount point for all containers in the build-pod. Tekton also
+adds it's own helper to ["git clone" the repository][ControllerGitClone] in place.
```

Comment:
It is not very related to your ship, but

> Tekton also adds it's own helper to ["git clone" the repository][ControllerGitClone] in place.

is mis-leading because Tekton is not doing much anymore when we `git clone` a repo.

Reply:
Indeed. Fixed on latest changes.